satellite/gc/service.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/auth/signing"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/uplink/piecestore"
+)
+
+var (
+	// Error defines the gc service errors class
+	Error = errs.Class("gc service error")
+)
+
+// Config contains configurable values for garbage collection
+type Config struct {
+	Interval time.Duration `help:"how frequently garbage collection filters should be sent to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Active   bool          `help:"set if garbage collection is actively running or not" releaseDefault:"true" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int64   `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log             *zap.Logger
+	config          Config
+	transport       transport.Client
+	overlay         overlay.DB
+	lastSendTime    time.Time
+	lastPieceCounts map[storj.NodeID]int
```

Comment:
Perhaps a topic for a separate PR, but I think we should keep track of the last piece counts in the database, so it can survive satellite restarts.

If one GC round is executed once per week, and the first one always initializes the bloom filters with 400,000 pieces then it would take at least 2 weeks to generate a bloom filter that will execute efficient garbage collection on storage nodes with millions of pieces.

It's very likely that during these 2 weeks the satellite will be restarted due to new deployment or another reason. If we don't keep this info in the DB then the bloom filter generation will start again with 400,000 pieces. This way big storage nodes may never get a good bloom filter.

Reply:
Could we just add a field to the node dossier or something? I think I agree, but I also think we should prioritize getting this merged first, then do the migration as a separate change

Reply:
This is a possible solution. I agree we can do this in a separate PR.

Reply:
https://storjlabs.atlassian.net/browse/V3-2157